TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,348 @@
  */
 #include "mbed.h"
 #include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
 #include "rtos.h"
 
+using namespace utest::v1;
+
+
 #if defined(MBED_RTOS_SINGLE_THREAD)
   #error [NOT_SUPPORTED] test not supported
 #endif
 
-#define TEST_STACK_SIZE 512
 
-#define SIGNAL_SET_VALUE    0x01
-const int SIGNALS_TO_EMIT = 100;
-const int SIGNAL_HANDLE_DELEY = 25;
+#define TEST_STACK_SIZE   512
+#define ALL_FLAGS 0x7fffffff
+#define PROHIBITED_FLAG 0x80000000
+#define MAX_FLAG_POS 30
+
+
+#define WAIT1()     sync_sem2.release(); \
+                    sync_sem1.wait()
+
+#define WAIT2()     sync_sem1.release(); \
+                    sync_sem2.wait()
+
+
+Semaphore sync_sem1(0, 1);
+Semaphore sync_sem2(0, 1);
+volatile int32_t ret_val;
+volatile osStatus ret_status;
+
+
+
+void run_set_clear(void)
+{
+    // 1.1
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    // 2.2
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    // 3.2
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    // 4.1
+    ret_val = Thread::signal_clr(ALL_FLAGS);
+    WAIT2();
+
+    // 4.3
+    ret_val = Thread::signal_clr(ALL_FLAGS);
+    WAIT2();
+
+    // 5.2
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    // 5.4
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    // 6.1
+    ret_val = Thread::signal_clr(ALL_FLAGS);
+    WAIT2();
+
+    // 6.3
+    ret_val = Thread::signal_clr(0);
+    WAIT2();
+
+    for (int i = 0; i <= MAX_FLAG_POS; i++) {
+        // 7.2
+        ret_val = Thread::signal_clr(0);
+        WAIT2();
+    }
+}
+
+template <int32_t signals, uint32_t millisec>
+void run_wait(void)
+{
+    osEvent ev = Thread::signal_wait(signals, millisec);
+    ret_status = ev.status;
+    sync_sem1.release();
+}
+
+template <int32_t signals, uint32_t millisec>
+void run_wait2(void)
+{
+    sync_sem1.release();
+    osEvent ev = Thread::signal_wait(signals, millisec);
+    ret_status = ev.status;
+    sync_sem1.release();
+}
+
+template <int32_t signals, uint32_t millisec>
+void run_wait3(void)
+{
+    sync_sem1.release();
+    sync_sem2.wait();
+    osEvent ev = Thread::signal_wait(signals, millisec);
+    ret_status = ev.status;
+    sync_sem1.release();
+}
+
+/** Test set and clear on no waiting thread
+
+    Given a thread
+    When call @a signal_set with specified signals
+    Then the specified signals are set
+    When call @a signal_clr with specified signals
+    Then the specified signals are cleared
+ */
+void test_set_clear(void)
```

Comment:
Can we not have such long tests? I think having to number and subnumber the flows means it's too complicated. Could we redesign it? Maybe split it up in separate test cases and use templates.